internal/config/config.go

```diff
@@ -0,0 +1,142 @@
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/project"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/resource"
+	"sigs.k8s.io/yaml"
+)
+
+const DefaultPath = "PROJECT"
+
+// Config is the unmarshalled representation of the configuration file
+type Config struct {
+	// Version is the project version, defaults to "1" (backwards compatibility)
+	Version string `json:"version,omitempty"`
+
+	// Domain is the domain associated with the project and used for API groups
+	Domain string `json:"domain,omitempty"`
+
+	// Repo is the go package name of the project root
+	Repo string `json:"repo,omitempty"`
+
+	// Resources tracks scaffolded resources in the project
+	// This info is tracked only in project with version 2
+	Resources []Resource `json:"resources,omitempty"`
+
+	// Multigroup tracks if the project has more than one group
+	MultiGroup bool `json:"multigroup,omitempty"`
+}
+
+// Check verifies that the configuration file exists
+func Check(path string) (bool, error) {
+	// Look up the file
+	_, err := os.Stat(path)
+
+	// If we could find it the file exists
+	if err == nil {
+		return true, nil
+	}
+
+	// Not existing and different errors are differentiated
+	if os.IsNotExist(err) {
+		err = nil
+	}
+	return false, err
+}
+
+func LoadFrom(path string) (*Config, error) {
+	// Read the file
+	in, err := ioutil.ReadFile(path) // nolint: gosec
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the file content
+	config := &Config{}
+	if err = yaml.Unmarshal(in, config); err != nil {
+		return nil, err
+	}
+
+	// kubebuilder v1 omitted version, so default to v1
+	if config.Version == "" {
+		config.Version = project.Version1
+	}
+
+	return config, nil
+}
+
+// SaveTo saves the configuration information to the provided path
+func (config Config) SaveTo(path string) error {
+	content, err := yaml.Marshal(config)
+	if err != nil {
+		return fmt.Errorf("error marshalling project configuration: %v", err)
+	}
+
+	err = ioutil.WriteFile(path, content, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to save configuration to %s: %v", path, err)
+	}
+
+	return nil
+}
+
+// ResourceGroups returns unique groups of scaffolded resources in the project
+func (config Config) ResourceGroups() []string {
+	groupSet := map[string]struct{}{}
+	for _, r := range config.Resources {
+		groupSet[r.Group] = struct{}{}
+	}
+
+	groups := []string{}
+	for g := range groupSet {
+		groups = append(groups, g)
+	}
+
+	return groups
+}
+
+// HasResource returns true if API resource is already tracked
+// NOTE: this works only for v2, since in v1 resources are not tracked
+func (config Config) HasResource(target *resource.Resource) bool {
+	for _, r := range config.Resources {
+		if r.IsEqualTo(target) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Resource contains information about scaffolded resources
+type Resource struct {
```

Comment:
Why Resource is defined in config.go?
Is this structure used outside from here?

Reply:
You are right, I will not export `config.Resource`

Reply:
I do nto know if YAML parsing will work with unexported types and fields, so I will rename it to `GVK` (from GroupVersionKind, the abbreviation is used in k8s quite often) to avoid having two `Resource` types.